rmanager.py

```diff
@@ -0,0 +1,46 @@
+class ResourceManager(object):
+    def __init__(self, ESstats, cpu_percent_usage, memory_percent_usage, njobs):
+        self.machineResources = { "total": {"cpu_75": cpu_percent_usage*int(ESstats["MachineCPUCount"]), 
+                                            "rss_75" : memory_percent_usage*int(ESstats["MachineMemoryGB"])*10737418 },
+                                  "available": {"cpu_75": cpu_percent_usage*int(ESstats["MachineCPUCount"]),
+                                            "rss_75" : memory_percent_usage*int(ESstats["MachineMemoryGB"])*10737418 } }
+        self.externalsStats = ESstats["externals"]# this should be get before running pkgtools
+        self.resourcesAllocatedForExternals = {} # resources that were alocated for given externals, say root -> "root": {"rss":"", cpu:""}
+        self.jobsOrderMetric = "cpu_75" # default to cpu
+        self.missingExternalsStrategy = None # runFirst, runLast
+    
+    def allocResourcesForExternals(self, externalsList=[]): # return ordered list for externals that can be started
+        # first, strip the names from build-*++version and make a tmp name to full name dict.
+        # toolfiles should not arrive here, but will also work if they do
+        ext_name_to_fullname_dict = {}
+        for e in externalsList:
+            short_name = e.split('+')[1]
+            ext_name_to_fullname_dict[short_name]=e
+        externalsList_shortNames = ext_name_to_fullname_dict.keys()
+        # if record for an external is not available, allocate it 1/4th of the resources and run it first
+        # OR allocate all resources for each missing external, forcing them to build last one by one.
+        externals_to_run = []
+        for ext in externalsList_shortNames:
+            if ext not in self.externalsStats:
+                # external is not found, this should happen only for new externals. get the first element whichever it is and change its properties
+                self.externalsStats[ext] = [{"name":ext, "cpu_75":self.machineResources["total"]["cpu_75"]/4,"rss_75":self.machineResources["total"]["rss_75"]/10 }]
```

Comment:
using 25% of resources for new/unknown packages is not safe. For example when we start a new release cycle or new architecture then all of the packages are unknown and in that case running 4 packages in parallel with `-j cpu_count` might overload the system. I would suggest to use 50% cpu and 50% rss and better to make these parameters configurable